src/transformers/models/vocos/modeling_vocos.py

```diff
@@ -0,0 +1,413 @@
+
+        Example:
+
+        ```python
+        >>> from datasets import load_dataset, Audio
+        >>> from transformers import VocosModel, VocosFeatureExtractor
+
+        >>> model = VocosModel.from_pretrained("Manel/Vocos")
+        >>> feature_extractor = VocosFeatureExtractor.from_pretrained("Manel/Vocos")
+
+        >>> ds = load_dataset("hf-internal-testing/librispeech_asr_dummy", "clean", split="validation")
+        >>> ds = ds.cast_column("audio", Audio(sampling_rate=feature_extractor.sampling_rate))
+        >>> audio_sample= ds[0]["audio"]["array"]
+
+        >>> inputs = feature_extractor(audio_sample, sampling_rate=feature_extractor.sampling_rate, return_tensors="pt")
+        >>> audio = model(inputs.input_features)
+        ```
+        """
+        hidden_states = self.backbone(features)
+        audio = self.head(hidden_states)
+        return audio
+
+
+@auto_docstring(
+    custom_intro="""
+    Vocos model with integrated EnCodec neural audio codec for end-to-end audio compression and reconstruction.
+    It accepts raw audio or pre-computed EnCodec RVQ codes (quantized audio features) and reconstructs a higher quality audio.
+    """
+)
+class VocosWithEncodecModel(VocosPreTrainedModel):
+    config_class = VocosWithEncodecConfig
+    main_input_name = "audio"
+
+    def __init__(self, config: VocosWithEncodecConfig):
+        super().__init__(config)
+
+        self.encodec_model = AutoModel.from_config(config.encodec_config)
+        self.num_quantizers = self.encodec_model.quantizer.get_num_quantizers_for_bandwidth(
+            bandwidth=max(config.bandwidths)
+        )
+        codebook_weights = torch.cat(
+            [layer.codebook.embed for layer in self.encodec_model.quantizer.layers[: self.num_quantizers]], dim=0
+        )
+        self.codebook_weights = nn.Parameter(codebook_weights, requires_grad=config.train_codebooks)
+        self.backbone = VocosBackbone(config)
+        self.head = VocosISTFTHead(config)
+        self.post_init()
+
+    def _audio_codes_to_features(self, codes: torch.LongTensor) -> torch.FloatTensor:
+        r"""
+        Transform a sequence of discrete tokens (codes) into feature embeddings using codebook weights.
+
+        Args:
+            codes (torch.LongTensor): Discrete codes of shape (n_codebooks, seq_len) or
+                (n_codebooks, batch_size, seq_len).
+
+        Returns:
+            torch.FloatTensor: Feature embeddings of shape (batch_size, hidden_dim, seq_len).
+        """
+        if codes.dim() == 2:
+            codes = codes.unsqueeze(1)
+        num_bins = self.encodec_model.quantizer.codebook_size
+        offsets = torch.arange(0, num_bins * codes.size(0), num_bins, device=codes.device).reshape(-1, 1, 1)
+        embeddings_idxs = codes + offsets
+        features = F.embedding(embeddings_idxs, self.codebook_weights).sum(dim=0)
+        return features.transpose(1, 2)
```

Comment:
this function would also be in the VocosProcessor so that features are directly outputted by the processor that can be fed into the model.

And bandwidth can also be an outputs of the processor so that it can be used by the model.